[DDP][Grad compression] Fix fp16 cpp hook

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63379
* __->__ #63375

I think tensor.copy_(tensor.to(torch::kFloat16)); will keep it as
float32.

Tested by add the following line:

```
LOG(INFO) << "Type is: " << compressed_tensor.scalar_type();
```

before:

```
I0816 17:03:09.823688 364141 default_comm_hooks.cpp:21] Type is: Float
```
after:

```
I0816 17:01:16.779052 353924 default_comm_hooks.cpp:21] Type is: Half
```

Differential Revision: [D30356256](https://our.internmc.facebook.com/intern/diff/D30356256/)

rohan-varma committed:
[DDP][Grad compression] Fix fp16 cpp hook

I think tensor.copy_(tensor.to(torch::kFloat16)); will keep it as
float32.

Tested by add the following line:

```
LOG(INFO) << "Type is: " << compressed_tensor.scalar_type();
```

before:

```
I0816 17:03:09.823688 364141 default_comm_hooks.cpp:21] Type is: Float
```
after:

```
I0816 17:01:16.779052 353924 default_comm_hooks.cpp:21] Type is: Half
```

Differential Revision: [D30356256](https://our.internmc.facebook.com/intern/diff/D30356256/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP][Grad compression] Fix fp16 cpp hook"


I think tensor.copy_(tensor.to(torch::kFloat16)); will keep it as
float32.

Tested by add the following line:

```
LOG(INFO) << "Type is: " << compressed_tensor.scalar_type();
```

before:

```
I0816 17:03:09.823688 364141 default_comm_hooks.cpp:21] Type is: Float
```
after:

```
I0816 17:01:16.779052 353924 default_comm_hooks.cpp:21] Type is: Half
```

Differential Revision: [D30356256](https://our.internmc.facebook.com/intern/diff/D30356256/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/c10d/default_comm_hooks.cpp:
```diff
@@ -16,21 +16,23 @@ c10::intrusive_ptr<c10::ivalue::Future> AllReduceCommHook::runHook(
     auto reduce_tensor = result.toTensorVector()[0];
-    reduce_tensor.copy_(reduce_tensor.to(torch::kFloat));
-    return c10::IValue(reduce_tensor);
+    const auto decompressed_tensor = bucket.getBuffer();
```
this more closely simulates `default_hooks.fp16_compress_hook` by restoring the type of the bucket's buffer.

SciPioneer replied:
Nit: const auto& will be better, to avoid copying tensor metadata.

SciPioneer commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -4112,6 +4116,18 @@ def _test_ddp_hook_parity(self, state, hook):
                     msg=f"Expected hook grad to be close to allreduce {avg[0, 0]}, but got {avg_hook[0, 0]}",
                 )
 
+        @sandcastle_skip_if(
```
No need to add this test. It's already covered by `test_builtin_ddp_comm_hooks_nccl`.

SciPioneer commented on torch/csrc/distributed/c10d/default_comm_hooks.cpp:
```diff
@@ -16,21 +16,23 @@ c10::intrusive_ptr<c10::ivalue::Future> AllReduceCommHook::runHook(
 
   auto allreduce_fut = state_->allreduce(tensors)->getFuture();
-  auto decompress = [](c10::ivalue::Future& allreduce_fut) {
+  auto decompress = [bucket](c10::ivalue::Future& allreduce_fut) {
```
bucket -> &bucket 
This can avoid copying vector data fields of `GradBucket`, such as `offsets`, `lengths`, `sizes_vec`. Copying these vectors can be non-trivial if there are a large number of parameters.

SciPioneer commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -4063,7 +4063,11 @@ def _test_ddp_hook_parity(self, state, hook):
                 process_group=process_group,
             )
-            net_with_hook.register_comm_hook(state=state, hook=hook)
+            try:
```
No need to update here if we don't need `test_ddp_hook_parity_allreduce_cpp_hook` below.

wayi1 reviewed (approved):
Thanks for the fix!

rohan-varma committed:
Update on "[DDP][Grad compression] Fix fp16 cpp hook"


I think tensor.copy_(tensor.to(torch::kFloat16)); will keep it as
float32.

Tested by add the following line:

```
LOG(INFO) << "Type is: " << compressed_tensor.scalar_type();
```

before:

```
I0816 17:03:09.823688 364141 default_comm_hooks.cpp:21] Type is: Float
```
after:

```
I0816 17:01:16.779052 353924 default_comm_hooks.cpp:21] Type is: Half
```

Differential Revision: [D30356256](https://our.internmc.facebook.com/intern/diff/D30356256/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP][Grad compression] Fix fp16 cpp hook"


I think tensor.copy_(tensor.to(torch::kFloat16)); will keep it as
float32.

Tested by add the following line:

```
LOG(INFO) << "Type is: " << compressed_tensor.scalar_type();
```

before:

```
I0816 17:03:09.823688 364141 default_comm_hooks.cpp:21] Type is: Float
```
after:

```
I0816 17:01:16.779052 353924 default_comm_hooks.cpp:21] Type is: Half
```

Differential Revision: [D30356256](https://our.internmc.facebook.com/intern/diff/D30356256/)

[ghstack-poisoned]
